src/envoy/http/service_control/filter_fuzz_test.cc

```diff
@@ -0,0 +1,222 @@
+#include "google/protobuf/text_format.h"
+#include "test/fuzz/fuzz_runner.h"
+#include "test/fuzz/utility.h"
+#include "test/mocks/http/mocks.h"
+#include "test/mocks/server/mocks.h"
+
+#include "api/envoy/http/service_control/config.pb.validate.h"
+#include "src/envoy/http/service_control/filter.h"
+#include "src/envoy/http/service_control/filter_config.h"
+#include "src/envoy/utils/filter_state_utils.h"
+#include "tests/fuzz/structured_inputs/service_control_filter.pb.validate.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+#include <fstream>
+#include <stdexcept>
+#include <string>
+
+namespace filter_api = ::google::api::envoy::http::service_control;
+namespace sc_api = ::google::api::servicecontrol::v1;
+using ::Envoy::Server::Configuration::MockFactoryContext;
+using ::testing::MockFunction;
+using ::testing::Return;
+using ::testing::ReturnRef;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace ServiceControl {
+namespace Fuzz {
+
+void doTest(ServiceControlFilter& filter, TestStreamInfo& stream_info,
+            const tests::fuzz::protos::ServiceControlFilterInput& input) {
+  // Decode headers.
+  bool end_stream = false;
+  auto downstream_headers =
+      Envoy::Fuzz::fromHeaders<Envoy::Http::TestRequestHeaderMapImpl>(
+          input.downstream_request().headers());
+  if (input.downstream_request().data().size() == 0 &&
+      !input.downstream_request().has_trailers()) {
+    end_stream = true;
+  }
+  filter.decodeHeaders(downstream_headers, end_stream);
```

Comment:
And then similarly, is there any reason to consider the filter status in case there's a reason to stop decoding?

Reply:
Not for this filter, we only ever `StopIteration` or `Continue`, and that will probably never change. Our filter should support all 3 being called.

Reply:
why do we need to make sure end_stream is correct?   Sometime, due to bug,  end_stream may not called correctly.  we still need to fuzz these data flow.

Reply:
Are you saying it's possible to execute `decodeHeaders(_, true)` and then `decodeData(_, _)` somewhere?

Reply:
@qiwzhang I think it's reasonable to expect `end_stream` is set correctly, this is part of the Envoy filter API. If it's set incorrectly, some upstream filters would not function. I don't think this will ever happen, but we can fuzz it if you believe it can.